src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs

```diff
@@ -430,15 +432,83 @@ private void RunTestCleanupMethod(object classInstance, TestResult result)
             }
             catch (Exception ex)
             {
-                var stackTraceInformation = StackTraceHelper.GetStackTraceInformation(ex.GetInnerExceptionOrDefault());
-                var errorMessage = string.Format(
-                    CultureInfo.CurrentCulture,
-                    Resource.UTA_CleanupMethodThrows,
-                    this.TestClassName,
-                    testCleanupMethod?.Name,
-                    StackTraceHelper.GetExceptionMessage(ex.GetInnerExceptionOrDefault()));
-                result.Outcome = TestTools.UnitTesting.UnitTestOutcome.Failed;
-                result.TestFailureException = new TestFailedException(UnitTestOutcome.Failed, errorMessage, stackTraceInformation);
+                var cleanupOutcome = UTF.UnitTestOutcome.Failed;
+                var cleanupError = new StringBuilder();
+                var cleanupStackTrace = new StringBuilder();
+                StackTraceInformation cleanupStackTraceInfo = null;
+
+                TestFailedException testFailureException = result.TestFailureException as TestFailedException;
+                if (testFailureException != null)
+                {
+                    if (!string.IsNullOrEmpty(testFailureException.Message))
+                    {
+                        cleanupError.Append(testFailureException.Message);
+                        cleanupError.AppendLine();
+                    }
+
+                    if (testFailureException.StackTraceInformation != null && !string.IsNullOrEmpty(testFailureException.StackTraceInformation.ErrorStackTrace))
+                    {
+                        cleanupStackTrace.Append(testFailureException.StackTraceInformation.ErrorStackTrace);
+                        cleanupStackTrace.Append(Environment.NewLine);
+                        cleanupStackTrace.Append(Resource.UTA_CleanupStackTrace);
+                        cleanupStackTrace.Append(Environment.NewLine);
+                    }
+                }
+
+                Exception realException = ex.GetInnerExceptionOrDefault();
+
+                // special case UnitTestAssertException to trim off part of the stack trace
+                if (realException is UnitTestAssertException)
```

Comment:
This looks like a common piece of code for a normal failing test. Can we reuse code?